src/ripple/app/tx/impl/TxQImpl.cpp

```diff
@@ -0,0 +1,969 @@
+
+    auto result = transactions_.emplace(std::make_pair(sequence, std::move(txn)));
+    assert(result.second);
+    assert(&result.first->second != &txn);
+
+    return result.first->second;
+}
+
+bool
+TxQAccount::removeCandidate(TxSeq const& sequence)
+{
+    return transactions_.erase(sequence) != 0;
+}
+
+CandidateTxn const*
+TxQAccount::findCandidateAt(TxSeq const& sequence) const
+{
+    auto iter = transactions_.find(sequence);
+    return iter == transactions_.end() ?
+        nullptr : &iter->second;
+}
+
+std::size_t
+FeeMetrics::updateFeeMetrics(Application& app,
+    ReadView const& view, bool timeLeap)
+{
+    std::vector<uint64_t> feeLevels;
+    for (auto const& tx : view.txs)
+    {
+        auto const baseFee = calculateBaseFee(app, view,
+            *tx.first, j_);
+        feeLevels.push_back(getFeeLevelPaid(*tx.first,
+            baseLevel, baseFee));
+    }
+    std::sort(feeLevels.begin(), feeLevels.end());
+    auto const size = feeLevels.size();
+
+    std::lock_guard <std::mutex> sl(lock_);
+
+    JLOG(j_.debug) << "Ledger " << view.info().seq <<
+        " has " << size << " transactions. " <<
+        "Ledgers are processing " <<
+        (timeLeap ? "slowly" : "as expected") <<
+        ". Expected transactions is currently " <<
+        txnsExpected_ << " and multiplier is " <<
+        escalationMultiplier_;
+
+    if (timeLeap)
+    {
+        // Ledgers are taking to long to process,
+        // so clamp down on limits.
+        txnsExpected_ = std::min(std::max(
+            feeLevels.size(), minimumTxnCount_),
+            targetTxnCount_ - 1);
+    }
+    else if (feeLevels.size() > txnsExpected_ ||
+        feeLevels.size() > targetTxnCount_)
+    {
+        // Ledgers are processing in a timely manner,
+        // so keep the limit high, but don't let it
+        // grow without bound.
+        txnsExpected_ = feeLevels.size();
+    }
+
+    if (feeLevels.empty())
+    {
+        escalationMultiplier_ = minimumMultiplier_;
+    }
+    else
+    {
+        // In the case of an odd number of elements, this
+        // evaluates to the middle element; for an even
+        // number of elements, it will add the two elements
+        // on either side of the "middle" and average them.
+        escalationMultiplier_ = (feeLevels[size / 2] +
+            feeLevels[(size - 1) / 2] + 1) / 2;
+        escalationMultiplier_ = std::max(escalationMultiplier_,
+            minimumMultiplier_);
+    }
+    JLOG(j_.debug) << "Expected transactions updated to " <<
+        txnsExpected_ << " and multiplier updated to " <<
+        escalationMultiplier_;
+
+    return size;
+}
+
+std::uint64_t
+FeeMetrics::scaleFeeLevel(OpenView const& view) const
+{
+    auto fee = baseLevel;
+
+    // Transactions in the open ledger so far
+    auto const current = view.txCount();
+
+    std::lock_guard <std::mutex> sl(lock_);
+
+    // Target number of transactions allowed
+    auto const target = txnsExpected_;
```

Comment:
We could write this lock free if we wanted. If we had a struct `FeeLevelInfo` with two 32-bit uints for `txnsExpected_` and `escalationMultiplier_`, and declared a member var of `std::atomic<FeeLevelInfo>` then that member var would likely be lock-free (it's implementation dependent, but I believe it would be likely). Then a simple `auto const fli = feeLevelInfo_.load()` gets us the info we need without a lock. I don't think I'd make the change now, but something to keep in mind.